### PR TITLE
Astonishing orchestra: use code to sign in

### PR DESCRIPTION
**Pull request by whimsicallyson**
This is, at the very least, the UI for signing in with the code provided in an email. currently I'm seeing CORS errors when trying to actually use a code... not sure if the issue is me, or the fact that its a remix.

https://glitch.manuscript.com/f/cases/3326594/login-update-email-sign-in-with-manual-sign-in-code-support-and-ui-to-match-editor

https://astonishing-orchestra.glitch.me/

### CODE REVIEW

File: src/presenters/pop-overs/sign-in-pop.jsx [hunk 1]
```diff
@@ -115,19 +179,31 @@ const SignInPopWithoutRouter = (props) => (
           hash: hash,
         },
       });
+      <CurrentUserConsumer>
+        {(currentUser, fetched, {login}) => <LoginPage setUser={login} destination={destination} setDestination={setDestination} {...props}/>}
+      </CurrentUserConsumer>;
```
Comment (GregWeil) [inline src/presenters/pop-overs/sign-in-pop.jsx]: The LoginPage is only meant to be used from the router, it'll take over the current page and redirect you somewhere else. You need to grab the login function that the LoginPage gets from the CurrentUserConsumer

File: src/presenters/pop-overs/sign-in-pop.jsx [hunk 2]
```diff
@@ -95,6 +97,68 @@ class EmailHandler extends React.Component {
...
+  onChange(e) {
+    this.setState({code: e.target.value});
+
+  }
+  
+  async onSubmit(e) {
+    e.preventDefault();
+    this.setState({done: true});
+    try {
+      await this.props.api.post('/login/email/?token=' + this.state.code);
```
Comment (GregWeil) [inline src/presenters/pop-overs/sign-in-pop.jsx]: This is where the login actually happens, so this is where that function from CurrentUserConsumer should get called.
```
const {data} = await this.props.api.post(...);
this.props.login(data);
```

File: src/presenters/pop-overs/sign-in-pop.jsx [hunk 3]
```diff
@@ -95,6 +96,69 @@ class EmailHandler extends React.Component {
...
+  onChange(e) {
+    this.setState({code: e.target.value});
+
+  }
+  
+  async onSubmit(e) {
+    e.preventDefault();
+    this.setState({done: true});
+    try {
+      const {data} = await this.props.api.post('/login/email/?token=' + this.state.code);
```
Comment (GregWeil) [inline src/presenters/pop-overs/sign-in-pop.jsx]: I think the issue is that your url here is wrong, this is the glitch.com url that the email sends you to, the api uses /auth/email/{code} (check out /login.jsx line 125)

Comment (whimsicallyson) [reply src/presenters/pop-overs/sign-in-pop.jsx]: good catch, thanks!

**Comment by whimsicallyson** [comment]
it works now! I feel confident about this!

**Comment by GregWeil** [comment]
It looks like the almost done/error in the email sent message is pulling some unwanted notification animations, it fades out unless you mouseover it. That's an issue on glitch.com too, so it doesn't have to block this review

**Pull request merged**